### ISSUE TITLE
huge huge memory usage!!

### ISSUE THREAD

**Issue by eisneim** [alxndrTL/mamba.py#9]
i find that the 1pscan1 method used in this Mamba implementation use huge amount of memory! any idea how to reduce memory consumption? or replace the pscan method with other implementation??

great thanks!

**Comment by alxndrTL** [comment]
Indeed, it is normal that this version uses a lot of memory as it doesn't use the recomputation technique dsescribed in the paper (see [this](https://github.com/alxndrTL/mamba.py?tab=readme-ov-file#performances) on the README for me information).
As of now, if you can use the default CUDA implementation, go for it, as mamba.py is mostly designed for educational purposes.
I'm working on a "performance update" and I hope that I will be able to include the recomputation technique.

Here is my benchmark for training a Mamba with `d_model=512`, `n_layers=16`, `B=16`, `L=512` on a A100 80GB :
- CUDA implementation : 3.2GB
- mamba.py : 33.6GB
You see how important the recomputation technique is.

**Comment by eisneim** [comment]
@alxndrTL thanks! this is such a great work!
i'm using M3 max macbook so CUDA implementation is not an option unfortunately; i was trying to use your mamba code with method described in paper vision mamba(Vim), it got the CLIP training working using mamba as text encoder, vision mamba as vision encoder, but comparing to Vit, the memory usage increased at least 5 times! 
maybe creating a custom operator using Metal would solve this issue? [apple doc](https://developer.apple.com/documentation/metal/metal_sample_code_library/customizing_a_pytorch_operation) 

looking forward to your "performance update"

**Comment by alxndrTL** [comment]
Yes, using Metal would be an option, but there is also MLX in the place now, I don't yet know if they are exclusive or not. I guess it will not be too long before MLX implements an efficient (kernel optimized, etc) pscan implementation.
I will keep you in touch here for the performance update!

**Comment by ali-shihab** [comment]
Hey, any updates on this performance/pscan update? My 16GB m1 pro just crashed from loading mamba-2.8b for fine tuning on a classification task, so I'm looking forward to this!

**Comment by alxndrTL** [comment]
Hello @ali-shihab, the [performance update](https://github.com/alxndrTL/mamba.py/pull/12) has been pushed ≈2 months ago, but it only enables faster training, not memory usage. I just worked on Jamba so I'm taking some time off but I'm aware that it's the main problem of mamba.py right now (and kind of make it not usable for real scenarios)

**Comment by ali-shihab** [comment]
Hey @alxndrTL, thanks for the quick response. If I'm not mistaken, the performance update is for PyTorch, no? Correct me if I'm wrong. 

Additionally, I've just scanned over jamba.py very quickly, and it seems everything in there can be implemented in MLX - do you think I could do this, or is there something that stopped you from being able to? If not, I'll see if I can implement a local version of it in MLX & clean it up for a PR once I have some time.

Also, that is completely understandable, enjoy your time off :)

**Comment by M-I** [comment]
flash attention is not there (yet):
https://github.com/ml-explore/mlx-examples/issues/724#issuecomment-2101093352
